vortex-array/src/arrays/primitive/compute/take.rs

```diff
@@ -1,67 +1,66 @@
+use std::mem::{MaybeUninit, transmute};
 use std::simd;
 
 use num_traits::AsPrimitive;
 use simd::num::SimdUint;
 use vortex_buffer::{Alignment, Buffer, BufferMut};
 use vortex_dtype::{
-    NativePType, Nullability, PType, match_each_integer_ptype, match_each_native_ptype,
-    match_each_native_simd_ptype, match_each_unsigned_integer_ptype,
+    DType, NativePType, PType, match_each_native_simd_ptype, match_each_unsigned_integer_ptype,
 };
-use vortex_error::VortexResult;
+use vortex_error::{VortexResult, vortex_bail};
 
 use crate::arrays::PrimitiveVTable;
 use crate::arrays::primitive::PrimitiveArray;
-use crate::compute::{TakeKernel, TakeKernelAdapter};
+use crate::compute::{TakeKernel, TakeKernelAdapter, cast};
 use crate::vtable::ValidityHelper;
 use crate::{Array, ArrayRef, IntoArray, ToCanonical, register_kernel};
 
 impl TakeKernel for PrimitiveVTable {
     #[allow(clippy::cognitive_complexity)]
     fn take(&self, array: &PrimitiveArray, indices: &dyn Array) -> VortexResult<ArrayRef> {
-        let indices = indices.to_primitive()?;
-
-        if array.ptype() != PType::F16
-            && indices.dtype().is_unsigned_int()
-            && indices.all_valid()?
-            && array.all_valid()?
-        {
-            // TODO(alex): handle nullable codes & values
-            match_each_unsigned_integer_ptype!(indices.ptype(), |C| {
+        let unsigned_indices = match indices.dtype() {
+            DType::Primitive(p, n) => {
+                if p.is_unsigned_int() {
+                    indices.to_primitive()?
+                } else {
+                    // This will fail if all values cannot be converted to unsigned
+                    cast(indices, &DType::Primitive(p.to_unsigned(), *n))?.to_primitive()?
+                }
+            }
+            _ => vortex_bail!("Invalid indices dtype: {}", indices.dtype()),
+        };
+
+        let validity = array.validity().take(unsigned_indices.as_ref())?;
+        if array.ptype() != PType::F16 {
+            match_each_unsigned_integer_ptype!(unsigned_indices.ptype(), |C| {
                 match_each_native_simd_ptype!(array.ptype(), |V| {
                     // SIMD types larger than the SIMD register size are beneficial for
                     // performance as this leads to better instruction level parallelism.
                     let decoded = take_primitive_simd::<C, V, 64>(
-                        indices.as_slice(),
+                        unsigned_indices.as_slice(),
```

Comment:
`unsigned_indices` might have undefined values (due to validity) that are out of bounds of the values (array)

Reply:
You have this problem already in the existing code. We assume all over that invalid values in indices are still valid take indices, however, I changed the simd implementation to do the checked gather so it will mask out out of bounds values